[sdk] Add get_transactions to verifying client. Extract common conversions to json-rpc-types.

phlip9 opened the pull request:
## Overview

The main goal of this PR is to add support for `get_transactions` to the verifying client. In order to accomplish this goal, we first need to do some type conversion refactoring and enrich the `get_transactions_with_proofs` API so that it reaches parity with the non-verifying `get_transactions` API.

## Key Questions

1. Do we need to make the new `include_events` parameter in `get_transactions_with_proofs` optional and default to false for backwards compatibility? OTOH, the API is "experimental", so maybe we just make a breaking change here. I don't have any particular opinion here : )

2. When extracting some conversions in `json-rpc::util`, there is one spot that depends on an unpublished diem-core crate: `move-explain`. Specifically, the `KeptVMStatus` -> `VMStatusView` conversion includes a human-readable explanation of a Move abort code, which is nice to have but relies on the unpublished `move-explain` crate. Do we hold this PR until we publish something that can convert abort codes to explanations or do we just leave the explanation blank for now?

## Commits

### [json-rpc] Add include_events param to get_transactions_with_proofs

This diff adds a previously missing parameter, namely, `include_events`, to the json-rpc request `get_transactions_with_proofs`. When this parameter is `true`, it includes all contract events associated with each transaction in the response.

It's important to add this functionality to `get_transactions_with_proofs` so that we can maintain parity with `get_transactions` for verifying clients.

Finally, we add a `serialized_events` field and `first_transaction_version` field to the `get_transactions_with_proofs` response. The `serialized_events` are an optional serialized list of events, which get included when `include_events` is true. The `first_transaction_version` field contains the version of the first transaction in the response, or `null` if there are no transactions.

Note: these events _are_ in fact verifiable. There was an incorrect comment previously that mentioned we could not verify the events, but this is not true.

### [json-rpc] extract json-rpc::util type conversions into views

We need these conversions accessible from inside `json-rpc-types::views` (they also belong there anyway) for the verifying client to be able to reuse the diem-types <-> json-rpc-types::views conversion logic needed to support `get_transactions`.

This diff is not particularly interesting, except for one caveat:

The move tx abort code -> human-readable explanation in `VMStatusView::from<&KeptVMStatus>` depends on logic in the `move-explain` crate getting published to crates.io

For now, this diff outputs a None/placeholder in until we get this crate published.

### [json-rpc] Fix mock get_transactions to return same # events as txs

The mock get_transactions used for testing json-rpc was returning too many event lists for the # of txs. For instance, if that current mock ledger version was 10, and I request 50 txs with events starting from version 5, the response should only include 5 txs and event lists.

Instead, the previous logic returned e.g. 5 txs and 50 event lists (since there are sometimes more event lists generated than actual txs for some reason). This diff fixes the logic so it will only ever return the same # of event lists as # actual txs returned.

### [json-rpc] Extract db <-> json conversions to views for get_transactions

To avoid code duplication and ensure that different json-rpc clients/servers convert json-rpc view types <-> diem core types equivalently, this diff extracts much of the conversion logic in get_transactions (and friends) into json-rpc-types views.

This change has the added benefit of increasing the testability of view types <-> core types conversion, since the IO logic (db lookup) is serparated from the type conversion logic (in views).

### [sdk] Add get_transactions to verifying client

Add support for JSON-RPC's get_transactions to the verifying client.  Now that the json-rpc view types <-> diem-core types conversion logic is extracted, the verifying client can reuse it to avoid duplication.

Include proptest for get_transactions. One notable limitation here is that we need to ensure both verifying and non-verifying clients avoid a race where start_version..start_version+limit contains more transactions in the second request, so we constrain the max limit to always give a valid response no matter the order/timing.

libra-action commented:
**This PR may have modified JSON-RPC server code.**

Breaking changes policy:

1. Chaning JSON-RPC method signatures or removing/modifying fields
   in the response in /V1 will not be accepted.
2. Adding fields is ok.
3. Adding a new JSON-RPC method is ok.

Please ensure you have also done the following to update the docs:

1. Update json-rpc/API-CHANGELOG.md.
2. Add/update type/method documentation under json-rpc/docs.
<!-- metadata: {"tag":"jsonrpc-compat"} -->

phlip9 commented:
1. Add method to get name of legacy stdlib script
2. Use (1.) to get transaction script name in Script -> ScriptView conversion, so we can use published sdk transaction-builder

bmwill commented on language/diem-framework/releases/artifacts/current/transaction_script_builder.rs:
```diff
@@ -3265,6 +3265,50 @@ impl ScriptCall {
             None => None,
         }
     }
+
```
IIRC this is a generated file so it might not be the best place for this. Maybe this could be added to the generator? or done in the transaction-builder crate itself?

phlip9 replied:
the change in `language/transaction-builder/generator/src/rust.rs` adds the codegen that generates this function : )

bmwill replied:
lol missed that :P

bmwill commented on json-rpc/types/src/views.rs:
```diff
@@ -634,6 +640,47 @@ impl std::fmt::Display for VMStatusView {
+            KeptVMStatus::Executed => VMStatusView::Executed,
+            KeptVMStatus::OutOfGas => VMStatusView::OutOfGas,
+            KeptVMStatus::MoveAbort(loc, abort_code) => {
+                // TODO(philiphayes): publish move-explain crate to crates.io
```
Also this is probably the part of the PR that I have the most concerns about/don't know a good solution for.

bmwill replied:
@vgao1996 Is this something that we could potentially package as part of the generated code or somehow make it more usable for the SDKs?

phlip9 replied:
My understanding is that in the same way we generate `language/diem-framework/releases/artifacts/current/transaction_script_builder.rs` we also generate `language/diem-framework/releases/artifacts/current/error_description/error_description.errmap`.

Currently, all the logic to read the file and explain error codes lives inside `language/move-prover/errmapgen/src/errmapgen.rs`; `move-explain` is just a thin CLI frontend. We can probably pull a similar trick as we did with the transaction builder: (1) `include_bytes!(<path-to-errmap>);`, (2) symlink to `errmapgen.rs`, and (3) add a `static ERROR_MAPPING: OnceCell<ErrorMapping>` that just `bcs::from_bytes` the static bytes.

bmwill replied:
I like it!

phlip9 replied:
cool : ) I'll see about putting out a quick PR with a new sdk crate for error code explanations

phlip9 replied:
https://github.com/diem/diem/pull/8410

phlip9 replied:
done! this PR now correctly gets the move error explanations : )

phlip9 commented on json-rpc/types/src/views.rs:
```diff
@@ -784,6 +877,85 @@ impl ScriptView {
+        // TODO(philiphayes): diem framework release onto crates.io
+        // let name = transaction_scripts::LegacyStdlibScript::try_from(script.code())
+        //     .map(|x| format!("{}", x))
+        //     .unwrap_or_else(|_| "unknown".to_string());
```
done!

xli commented on json-rpc/docs/method_get_transactions_with_proofs.md:
```diff
@@ -2,29 +2,34 @@
+|---------------------------|------------------------|-------------------------------------------------------------------------------------------------------------|
+| serialized_transactions   | List<string>           | An array of hex encoded strings with the raw bytes of the returned `Transaction`                            |
+| serialized_events         | Option<string>         | An optional BCS-serialized list of each transaction's list of events or `null` if `include_events` is false |
+| first_transaction_version | Option<unsigned int64> | If there are any transactions, then the version of the first transaction, otherwise `null`                  |
```
why we need this? isn't the `start_version` in the parameter is the first transaction version?

phlip9 replied:
I could go either way. On the one hand, I'd prefer the *_with_proofs APIs to avoid trying to be smart and instead just expose the underlying storage API without changes. On the other hand, we can save some bytes and remove this field which the caller can reconstitute. I definitely lean toward "no changes", but if other people have stronger opinions I can remove this field.

xli replied:
I think underline APIs are free to change, json-rpc API response should be more restricted to make sure it can be as stable as possible.
For this field, the question is not what it is, the question is why it is different with the `start` param. If they are never different, why we output it.

phlip9 replied:
Makes sense, I'll remove this field 👍

xli commented on json-rpc/tests/integration_test.rs:
```diff
@@ -548,12 +548,14 @@ fn create_test_cases() -> Vec<Test> {
-                            "reason_description": " The account does not hold a large enough balance in the specified currency"
-                        },
+                        "explanation": serde_json::Value::Null,
+                        // TODO(philiphayes): re-enable when Move error explanations are published
```
I don't think leave a todo is good idea

bmwill replied:
Philip is leaving the TODOs in the code in order to drive comments during Review, I don't expect this PR to land with these TODOs still here.

xli replied:
maybe use TBD, or discussion?

phlip9 replied:
re-enabled previous expected output now that explanations are working again

xli commented on json-rpc/types/src/request.rs:
```diff
@@ -319,6 +319,8 @@ pub struct GetAccountStateWithProofParams {
 pub struct GetTransactionsWithProofsParams {
     pub start_version: u64,
     pub limit: u64,
+    // TODO(philiphayes): do we need default=false for backwards compatibility?
+    pub include_events: bool,
```
I think this field should be optional, not because BC, but because it is an option.

bmwill replied:
counter example is that the corresponding non-verifying api doesn't have this param as optional

xli replied:
I think non-verifying API made mistake to enforce it as required parameter. And probably because at the time we create the API, there is no optional for creating optional parameter.

bmwill replied:
I still don't have strong feelings but if we do want to default it to false then that should be a simple annotation.

xli commented on json-rpc/types/src/views.rs:
```diff
@@ -644,18 +692,183 @@ pub struct TransactionView {
+            transactions.len(),
+        );
+
+        let events = if let Some(events) = txs.events {
```
maybe should name this events_list, "events" is confusing as one txn has a list of events

xli commented on json-rpc/types/src/views.rs:
```diff
@@ -644,18 +692,183 @@ pub struct TransactionView {
 pub struct TransactionsWithProofsView {
+    // TODO(philiphayes): just serialize as a giant BytesView instead of a Vec<_>?
     pub serialized_transactions: Vec<BytesView>,
+    pub serialized_events: Option<BytesView>,
```
add #[serde(skip_serializing_if = "Option::is_none")], same for first_transaction_version

xli commented on sdk/client/src/verifying_client.rs:
```diff
@@ -421,69 +439,139 @@ fn verifying_get_account(address: AccountAddress, version: Option<Version>) -> V
+        // list _all_ BalanceResource's under balance? Then we could avoid
+        // this extra lookup. It's even more problematic when a batch
+        // includes many GetAccount requests, as this batch handling logic
+        // isn't (currently) smart enough to dedup the diem_root lookups.
```
wondering what is this todo for

phlip9 replied:
ahh, if you look at that commit, it's just a reformatting commit (changing the indentation).

as for the comment, it's about more efficiently extracting the balance resources from an account without having to first look up the global supported currencies in DiemRoot.

xli replied:
So, sometime, you do leave a TODO :D

phlip9 replied:
#8415 takes care of this TODO :^)

xli commented on json-rpc/src/tests/utils.rs:
```diff
@@ -148,9 +148,10 @@ impl DbReader for MockDiemDB {
+        ledger_version: u64,
         fetch_events: bool,
     ) -> Result<TransactionListWithProof, Error> {
+        let limit = std::cmp::min(limit, ledger_version - start_version + 1);
```
Shouldn't we raise error if start_version + limit > ledger_version?

phlip9 replied:
Why would we return an error? If I ask for the next 1000 txs, but there are only 500 new txs, should we throw an error or just return the 500 new txs? The `get_transactions_with_proofs` doesn't let the user specify the `ledger_version`, so the user doesn't know  the current ledger version (which would define the max limit).

xli replied:
I'd expect JSON-RPC API handler will do input param validation, thus all parameters passed into DB layer should be safe.
Thus in this test utility code, we should assert the parameters are safe, not fix them.

phlip9 replied:
Hmm, I'd actually go the other way, json-rpc is "just" another client of diemdb, so I wouldn't expect all db clients to need to repeat the same input validation, since db already correctly handles input validation in this case. For example, we can remove the `if start_version > ledger_version || limit == 0 { return Ok(TransactionListView::empty()) }`

phlip9 replied:
For example, you can see here the `get_transactions` handler becomes super simple:

```rust
/// Returns transactions by range
pub fn get_transactions(
    db: &dyn DbReader,
    ledger_version: u64,
    start_version: u64,
    limit: u64,
    include_events: bool,
) -> Result<TransactionListView, JsonRpcError> {
    let txs = db.get_transactions(start_version, limit, ledger_version, include_events)?;
    Ok(TransactionListView::try_from(txs)?)
}
```

xli replied:
1. The error raised from DB is not well defined. If you want db handles all of these, you need make sure db interface does what you described before you make the change. 
2. You can keep adding layer for sharing logic, but somewhere need handle these logic. Current method does all input validation and response proper jsonrpc error. 
3. I don't really care where we put these logic, but it should be consistent.

xli replied:
actually, there is layer created for this purpose already, it is data.rs in same directory with method.rs.

phlip9 replied:
I see what you're saying. The DB layer probably needs some work to be more consistent with its input validation and error handling. The interface stability is also a good point. In lieu of improving the db interface, I've added the checking back and some assertions in the mock db to make sure inputs are validated.

xli commented on json-rpc/src/util.rs:
```diff
@@ -120,160 +110,6 @@ impl SdkInfo {
     }
 }
 
-pub fn vm_status_view_from_kept_vm_status(status: &KeptVMStatus) -> VMStatusView {
```
it would be easier for every one if you could just submit a PR to refactor these logic (moving the logic around without changing anything).
in current format, I can't really review these change.

bmwill replied:
@xli Philip does a really good job at breaking down PRs into individual atomic commits. Please use the "commit" view when reviewing these changes. It makes reviewing these changes much easier to read.

phlip9 replied:
Looks like github doesn't do a very good job of displaying the code move. FWIW, in that diff I didn't change anything except move the function bodies and comment out the move-explain part.

xli replied:
right, I forgot to use commits....

xli commented on json-rpc/types/src/views.rs:
```diff
@@ -644,18 +644,60 @@ pub struct TransactionView {
+
 #[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
 pub struct TransactionsWithProofsView {
+    // TODO(philiphayes): just serialize as a giant BytesView instead of a Vec<_>?
```
I think the question is whether you need create a new type for serializing and deserializing. If you need, it is better to just define the type needed in json-rpc response. Otherwise, maybe we should create BCS-RPC :D

phlip9 replied:
going to just leave this as a `Vec<BytesView>` for now

phlip9 commented:
rebased onto main, which includes the move error explanations inside sdk/transaction-builder. after this rebase, we no longer have any issues with modified output behaviour! i've also addressed comments in the fixup commits

bors-libra commented:
:exclamation: Invalid command

<details>
<summary>Bors help and documentation</summary>
<br />

Bors is a Github App used to manage merging of PRs in order to ensure that CI is always green. It does so by maintaining a [Merge Queue](https://github.com/diem/diem/projects/2). Once a PR reaches the head of the Merge Queue it is rebased on top of the latest version of the PR's `base-branch` (generally `master`) and then triggers CI. If CI comes back green the PR is then merged into the `base-branch`. Regardless of the outcome, the next PR is the queue is then processed.

### General
- This project's Merge Queue can be found [here](https://github.com/diem/diem/projects/2).
- This project requires PRs to be [Reviewed](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-request-reviews) and Approved before they can be queued for merging.
- Before PR's can be merged they must be configured with ["Allow edits from maintainers"](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) enabled. This is needed for Bors to be able to update PR's in-place so that Github can properly recognize and mark them as "merged" once they've been merged into the upstream branch.
### Commands
Bors actions can be triggered by posting a comment which includes a line of the form `/<action>`.
| Command | Action | Description |
| --- | --- | --- |
| __Land__ | `land`, `merge` | attempt to land or merge a PR |
| __Canary__ | `canary`, `try` | canary a PR by performing all checks without merging |
| __Cancel__ | `cancel`, `stop` | stop an in-progress land |
| __Cherry Pick__ | `cherry-pick <target>` | cherry-pick a PR into `<target>` branch |
| __Priority__ | `priority` | set the priority level for a PR (`high`, `normal`, `low`) |
| __Help__ | `help`, `h` | show this help message |

### Options
Options for Pull Requests are configured through the application of labels.
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Option&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; | Description |
| --- | --- |
| ![label: bors-high-priority](https://img.shields.io/static/v1?label=&message=bors-high-priority&color=lightgrey) | Indicates that the PR is high-priority. When queued the PR will be placed at the head of the merge queue. |
| ![label: bors-low-priority](https://img.shields.io/static/v1?label=&message=bors-low-priority&color=lightgrey) | Indicates that the PR is low-priority. When queued the PR will be placed at the back of the merge queue. |
| ![label: bors-squash](https://img.shields.io/static/v1?label=&message=bors-squash&color=lightgrey) | Before merging the PR will be squashed down to a single commit, only retaining the commit message of the first commit in the PR. |

</details>

phlip9 commented:
rebased

bmwill reviewed (approved):
The changes you made to move the error_description stuff into the sdk really made all my concerns with this PR go away. Thanks for doing that.

There's only really 1 nit that Xiao brought up that should be easy to address.

phlip9 commented:
squash and rebase

/land

bors-libra commented:
:exclamation: Land has been canceled due to this PR being updated with new commits. Please issue another Land command if you want to requeue this PR.

phlip9 commented:
had to fix some fixups that I applied to the wrong commit D:

/land
